activation/activation.go

```diff
@@ -243,30 +243,28 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) (bool, error) {
 	if err != nil && !posEpoch.IsGenesis() {
 		return false, fmt.Errorf("failed to get active set size: %v", err)
 	}
-	view, err := b.mesh.GetOrphanBlocksBefore(b.mesh.LatestLayer())
-	if err != nil {
-		return false, fmt.Errorf("failed to get current view: %v", err)
+
+	viewLayer := b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch).FirstLayer(b.layersPerEpoch)
+	var view []types.BlockID
+	if viewLayer > 0 {
+		view, err = b.mesh.GetOrphanBlocksBefore(viewLayer)
+		if err != nil {
+			return false, fmt.Errorf("failed to get current view: %v", err)
+		}
 	}
 	atx := types.NewActivationTxWithChallenge(*b.challenge, b.coinbaseAccount, activeIds, view, b.nipst)
-	activeSetSize, err := b.db.CalcActiveSetFromView(atx) // TODO: remove this assertion to improve performance
-	if err != nil && !atx.TargetEpoch(b.layersPerEpoch).IsGenesis() {
-		return false, fmt.Errorf("failed to calc active set from view: %v", err)
-	}
-	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
-		log.Uint32("cache_cnt", activeIds), log.Uint32("view_cnt", activeSetSize))
 
-	if !atx.TargetEpoch(b.layersPerEpoch).IsGenesis() && activeSetSize == 0 {
-		b.log.Warning("empty active set size found! len(view): %d, view: %v", len(atx.View), atx.View)
-		return false, nil
+	viewAsBytes, err := types.ViewAsBytes(view)
+	if err != nil {
+		return false, fmt.Errorf("failed to serialize view: %v", err)
 	}
+	activesetCache.Add(common.BytesToHash(viewAsBytes), activeIds)
+
+	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
+		log.Uint32("cache_cnt", activeIds))
 
 	if atx.TargetEpoch(b.layersPerEpoch).IsGenesis() {
 		atx.ActiveSetSize = 0
```

Comment:
We may want to change this to "GenesisActiveSet" from Gavrad's PR.

There's an open issue that the validating side does not check this condition (i.e. does not check if the target epoch is genesis and active set is zero); we had some issues about it with nodes that join late.

Reply:
Sounds reasonable, but I think it's out of scope of this PR.